sympy/ntheory/residue_ntheory.py

```diff
@@ -1392,3 +1401,61 @@ def quadratic_congruence(a, b, c, p):
         for j in root:
             res.add(j % p)
     return sorted(res)
+
+
+def _general_congruence_prime(coefficients, p):
+    roots = []
+    rank = len(coefficients)
+    for i in range(0, p):
+        f_val = 0
+        for coeff in range(0,rank - 1):
+            f_val = (f_val + pow(i, int(rank - coeff - 1), p) * coefficients[coeff]) % p
+        f_val = f_val + coefficients[-1]
+        if f_val % p == 0:
+            roots.append(i)
+    return roots
+
+
+def _diff_poly(root, coefficients, p):
+
+    diff = 0
+    rank = len(coefficients)
+    for coeff in range(0,rank - 1):
+        if not coefficients[coeff]:
+            continue
+        diff = (diff + pow(root, rank - coeff - 2, p) * (rank - coeff - 1) * coefficients[coeff]) % p
+    return diff % p
+
+
+def _val_poly(root, coefficients, p):
+    rank = len(coefficients)
+    f_val = 0
+    for coeff in range(0, rank - 1):
+        f_val = (f_val + pow(root, rank - coeff - 1, p) * coefficients[coeff]) % p
+    f_val = f_val + coefficients[-1]
+    return f_val % p
+
+
+def general_congruence(expr, m):
+    """
+    Find the solutions to a polynomial congruence equation.
+    """
+
+    from sympy import Poly
+    if not expr.is_polynomial():
+        raise ValueError("The expression should be a polynomial")
+    syms = expr.free_symbols
+    if len(syms) > 1:
+        raise ValueError("Do not support for more than one symbol")
+    coefficients = Poly(expr,syms).all_coeffs()
+    rank = len(coefficients)
+    if rank == 3:
+        return quadratic_congruence(coefficients[0], coefficients[1], coefficients[2], m)
+    if rank == 2:
+        return quadratic_congruence(0, coefficients[0], coefficients[1], m)
+    if isprime(m):
+        return _general_congruence_prime(coefficients, m)
+    return _help(m,
```

Comment:
NIce! Doesn't that just make you happy? :-)

Did the tests pass? If so, perhaps you can think of a better name for what that algorithm is doing and then add a little docstring indicating what it is doing. Even better names than `how1:3` could be useful...it was just a suggestion-of-concept that I was giving and you got the gist.

Reply:
Yes, the code looks more organized now. I have added more test cases to increase test coverage.

Reply:
I don't know why the check is failing.

Reply:
I got it.  in #18295 there is an unused import and it was committed.

Reply:
I have changed the names.